From b8bb3efebab7c20fc039b43b53436840feb78162 Mon Sep 17 00:00:00 2001
From: Boris Brezillon <boris.brezillon@free-electrons.com>
Date: Fri, 8 Jul 2016 20:32:47 +0200
Subject: [PATCH] nand_util: write page per page

Signed-off-by: Boris Brezillon <boris.brezillon@free-electrons.com>
---
 drivers/mtd/nand/nand_util.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/mtd/nand/nand_util.c b/drivers/mtd/nand/nand_util.c
index 71285b666941..248169af370b 100644
--- a/drivers/mtd/nand/nand_util.c
+++ b/drivers/mtd/nand/nand_util.c
@@ -645,10 +645,10 @@ int nand_write_skip_bad(nand_info_t *nand, loff_t offset, size_t *length,
 			continue;
 		}
 
-		if (left_to_write < (blocksize - block_offset))
+		if (left_to_write < nand->writesize)
 			write_size = left_to_write;
 		else
-			write_size = blocksize - block_offset;
+			write_size = nand->writesize;
 
 		truncated_write_size = write_size;
 #ifdef CONFIG_CMD_NAND_TRIMFFS
-- 
2.7.4

